src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.Sender.cs

```diff
@@ -0,0 +1,258 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Net.WebSockets.Compression;
+using System.Security.Cryptography;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.WebSockets
+{
+    internal partial class ManagedWebSocket
+    {
+        private sealed class Sender : IDisposable
+        {
+            private readonly int _maskLength;
+            private readonly WebSocketDeflater? _deflater;
+            private readonly Stream _stream;
+
+            private readonly Buffer _buffer = new();
+
+            public Sender(Stream stream, WebSocketCreationOptions options)
+            {
+                _maskLength = options.IsServer ? 0 : MaskLength;
+                _stream = stream;
+
+                var deflate = options.DeflateOptions;
+
+                if (deflate is not null)
+                {
+                    // If we are the server we must use the client options
+                    _deflater = options.IsServer ?
+                        new WebSocketDeflater(deflate.ClientMaxWindowBits, deflate.ClientContextTakeover) :
+                        new WebSocketDeflater(deflate.ServerMaxWindowBits, deflate.ServerContextTakeover);
+                }
+            }
+
+            public void Dispose() => _deflater?.Dispose();
+
+            public ValueTask SendAsync(MessageOpcode opcode, bool endOfMessage, ReadOnlyMemory<byte> content, CancellationToken cancellationToken = default)
+            {
+                bool compressed = false;
+
+                // Encoding is onlt supported for user messages
+                if (_deflater is not null && opcode <= MessageOpcode.Binary)
+                {
+                    _buffer.EnsureFreeCapacity(MaxMessageHeaderLength + (int)(content.Length * 0.6));
```

Comment:
To provoke a discussion about it :) 

I will need to profile the code and decide on a good strategy to guess potentially how much compression ratio are we expecting depending on the different settings used. Or may be instead of using a single continuous buffer, which is copied when more space is needed, we could have a list of buffers (think ReadOnlySequence<byte>).

The more I think of it, the more I find good reasons to do it. It would complicate the buffer management in the sender little bit, but will avoid the need to resize arrays and memcopy data. Before doing any perf optimizations though, I need guidance how to write perf tests that live inside the current source or at least to be able to use it.